### PR TITLE
feat(router): expose router operation cache metrics

### PR DESCRIPTION
**Pull request by Noroth**
## Motivation and Context

Add an option to expose OTEL metrics from `ristretto caches`.

## Example Dashboard

image

## Checklist

- [ ] I have discussed my proposed changes in an issue and have received approval to proceed.
- [ ] I have followed the coding standards of the project.
- [ ] Tests or benchmarks have been added or updated.
- [ ] Documentation has been updated on [https://github.com/wundergraph/cosmo-docs](https://github.com/wundergraph/cosmo-docs).
- [ ] I have read the [Contributors Guide](https://github.com/wundergraph/cosmo/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: router/pkg/metric/cache_metrics.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+}
+
+func (c *CacheMetrics) observeForCacheType(o otelmetric.Observer, cacheType string, metrics *ristretto.Metrics, maxCost int64) {
+	if metrics == nil {
+		return
+	}
+
+	o.ObserveInt64(cacheHitStats, int64(metrics.Hits()),
+		otelmetric.WithAttributes(c.baseAttributes...),
+		otelmetric.WithAttributes(attribute.String("type", "hits")),
```
Comment (jensneuse) [inline router/pkg/metric/cache_metrics.go:144]: Would it make sense to put literals like "type", "hits", "misses", "cache_type", etc. into const strings?

File: router/pkg/metric/cache_metrics.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+		operationCacheCostMaxMetric,
+		otelmetric.WithDescription("Tracks the maximum configured cost for a cache. Useful to investigate differences between the number of keys and the current cost"),
+	); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *CacheMetrics) Observe(
```
Comment (jensneuse) [inline router/pkg/metric/cache_metrics.go:110]: the func name Observe indicates that this func will be blocking. When looking closer, it appears that this is not the case. Maybe different naming could clarify this?

Alternatively, you could add comments to the functions to make the functionality clear.

**Review by jensneuse — APPROVED** [review]
Please check with @StarpTech but overall looks good to me except some minor things.

File: router/core/graph_server.go [hunk 3]
```diff
@@ -453,6 +480,37 @@ func (s *graphServer) buildGraphMux(ctx context.Context,
 		}
 	}
 
+	if s.cacheMetrics != nil {
```
Comment (StarpTech) [inline router/core/graph_server.go:483]: Shutdown when the mux is shutdown

File: router-tests/prometheus_test.go [hunk 4]
```diff
@@ -2956,6 +2977,188 @@ func TestPrometheus(t *testing.T) {
 			require.Greater(t, len(requestsInFlightMetricsFull[1].Label), len(requestsInFlightMetricsFiltered[1].Label))
 		})
 	})
+
+	t.Run("Collect router cache metrics", func(t *testing.T) {
```
Comment (StarpTech) [inline router-tests/prometheus_test.go:2981]: Please separate tests for telemetry and cache details

**Review by StarpTech — CHANGES_REQUESTED** [review]

File: router-tests/prometheus_test.go [hunk 5]
```diff
@@ -2956,6 +2977,189 @@ func TestPrometheus(t *testing.T) {
...
+			})
+
+			require.JSONEq(t, employeesTagData, res.Body)
+
+			metricFamilies, err = promRegistryFiltered.Gather()
+			require.NoError(t, err)
+
+			cacheMetrics := findCacheMetrics(metricFamilies)
+
+			for _, c := range cacheMetrics {
```
Comment (StarpTech) [inline router-tests/prometheus_test.go:3031]: Can we simplify this and make assertions more readable?

File: router/pkg/metric/cache_metrics.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+		otelmetric.WithAttributes(c.baseAttributes...),
+		otelmetric.WithAttributes(
+			otel.CacheMetricsOperationAttribute.String("added"),
+			otel.CacheMetricsCacheTypeAttribute.String(cacheType)),
+	)
+
+	o.ObserveInt64(cacheCostStats, int64(metrics.CostEvicted()),
+		otelmetric.WithAttributes(c.baseAttributes...),
+		otelmetric.WithAttributes(
+			otel.CacheMetricsOperationAttribute.String("evicted"),
```
Comment (StarpTech) [inline router/pkg/metric/cache_metrics.go:188]: We could also move those const to attributes.go. WDYT?

**Review by StarpTech — APPROVED** [review]
LGTM

**Pull request merged**